pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy

```diff
@@ -200,19 +211,96 @@ public class Utils {
+                r.addAction(new ExecutionModelAction(stages))
+            }
+
+            if (root != null) {
+                root = populateWhen(root, model)
+            }
+        }
+
+        return root
+    }
+
+    /**
+     * Attaches the {@link StageConditionals} to the appropriate {@link Stage}s, pulling from the AST model.
+     *
+     * @param root
+     * @param model
+     * @return an updated {@link Root}
+     */
+    static Root populateWhen(@Nonnull Root root, @Nonnull ModelASTPipelineDef model) {
+        List<Stage> stagesWithWhen = []
+
+        root.stages.stages.each { s ->
+            ModelASTStage astStage = model.stages.stages.find { it.name == s.name }
+            if (astStage.when != null) {
+                s.when(new StageConditionals(stageConditionalFromAST(astStage.when.condition)))
+            }
+            stagesWithWhen.add(s)
+        }
+
+        root.stages.stages = stagesWithWhen
+
+        return root
+    }
+
+    /**
+     * Translates the {@link ModelASTWhenContent} into a {@link DeclarativeStageConditional}.
+     *
+     * @param w
+     * @return A populated {@link DeclarativeStageConditional}
+     */
+    private static DeclarativeStageConditional stageConditionalFromAST(ModelASTWhenContent w) {
+        DeclarativeStageConditional c = null
+        DeclarativeStageConditionalDescriptor desc = DeclarativeStageConditionalDescriptor.byName(w.name)
+
+        if (w instanceof ModelASTWhenCondition) {
+            if (desc.allowedChildrenCount == 0) {
+                Object[] arg = new Object[1]
+                arg[0] = w.args?.argListToMap()
+                c = (DeclarativeStageConditional)getDescribable(w.name, desc.clazz, arg).instantiate()
+            } else if (desc.allowedChildrenCount == 1) {
+                DeclarativeStageConditional single = stageConditionalFromAST(w.children.first())
+                c = (DeclarativeStageConditional)getDescribable(w.name, desc.clazz, single).instantiate()
+            } else {
+                List<DeclarativeStageConditional> nested = w.children.collect { stageConditionalFromAST(it) }
+                c = (DeclarativeStageConditional)getDescribable(w.name, desc.clazz, nested).instantiate()
+            }
+        } else if (w instanceof ModelASTWhenExpression) {
+            ModelASTWhenExpression expr = (ModelASTWhenExpression)w
+
+            c = (DeclarativeStageConditional)getDescribable(w.name, desc.clazz, expr.codeBlockAsString()).instantiate()
         }
-        if (r.getAction(ExecutionModelAction.class) == null) {
-            r.addAction(new ExecutionModelAction(stages))
+
+        return c
+    }
+
+    /**
+     * Takes a string and makes sure it starts/ends with double quotes so that it can be evaluated correctly.
+     *
+     * @param s The original string
+     * @return Either the original string, if it already starts/ends with double quotes, or the original string
+     * prepended/appended with double quotes.
```

Comment:
So actually, I don't think it does. We're just taking the string from the original Jenkinsfile source, in effect.